mongodbatlas.go

```diff
@@ -59,37 +62,68 @@ func Run(apiTLSConfig *api.TLSConfig) error {
 	return nil
 }
 
-func (m *MongoDBAtlas) CreateUser(ctx context.Context, statements dbplugin.Statements, usernameConfig dbplugin.UsernameConfig, expiration time.Time) (username string, password string, err error) {
-	// Grab the lock
+func (m *MongoDBAtlas) Initialize(ctx context.Context, req newdbplugin.InitializeRequest) (newdbplugin.InitializeResponse, error) {
 	m.Lock()
 	defer m.Unlock()
 
-	statements = dbutil.StatementCompatibilityHelper(statements)
+	m.RawConfig = req.Config
+
+	err := mapstructure.WeakDecode(req.Config, m.mongoDBAtlasConnectionProducer)
+	if err != nil {
+		return newdbplugin.InitializeResponse{}, err
+	}
+
+	if len(m.PublicKey) == 0 {
+		return newdbplugin.InitializeResponse{}, errors.New("public Key is not set")
+	}
+
+	if len(m.PrivateKey) == 0 {
+		return newdbplugin.InitializeResponse{}, errors.New("private Key is not set")
+	}
+
+	// Set initialized to true at this point since all fields are set,
+	// and the connection can be established at a later time.
+	m.Initialized = true
 
-	if len(statements.Creation) == 0 {
-		return "", "", dbutil.ErrEmptyCreationStatement
+	resp := newdbplugin.InitializeResponse{
+		Config: req.Config,
+	}
+
+	return resp, nil
+}
+
+func (m *MongoDBAtlas) NewUser(ctx context.Context, req newdbplugin.NewUserRequest) (newdbplugin.NewUserResponse, error) {
+	// Grab the lock
+	m.Lock()
+	defer m.Unlock()
+
+	if len(req.Statements.Commands) == 0 {
+		return newdbplugin.NewUserResponse{}, dbutil.ErrEmptyCreationStatement
 	}
 
 	client, err := m.getConnection(ctx)
 	if err != nil {
-		return "", "", err
+		return newdbplugin.NewUserResponse{}, err
 	}
 
-	username, err = m.GenerateUsername(usernameConfig)
+	username, err := m.GenerateUsername(dbplugin.UsernameConfig{
+		DisplayName: req.UsernameConfig.DisplayName,
+		RoleName:    req.UsernameConfig.RoleName,
+	})
 	if err != nil {
-		return "", "", err
+		return newdbplugin.NewUserResponse{}, err
 	}
 
-	password, err = m.GeneratePassword()
+	password, err := m.GeneratePassword()
 	if err != nil {
-		return "", "", err
+		return newdbplugin.NewUserResponse{}, err
 	}
 
 	// Unmarshal statements.CreationStatements into mongodbRoles
 	var databaseUser mongoDBAtlasStatement
-	err = json.Unmarshal([]byte(statements.Creation[0]), &databaseUser)
+	err = json.Unmarshal([]byte(req.Statements.Commands[0]), &databaseUser)
```

Comment:
Since any statements beyond the first one are ignored, can you return an error if there are more than one?

Reply:
It seems that we are currently ignoring the other statements if they are provided (same for the MongoDB implementation/update). Would this be a breaking behavior if this is changed?

Reply:
Technically it would since we'd be erroring if they provide more than one command rather than ignoring them, however I think this a bad user experience if we leave it as-is since we claim that we'll do something (additional commands) but don't actually.

Reply:
Theron pointed me to the elasticsearch bit of code where we do this. I think it's a fair point, though we're should probably do the same for the mongodb (non-Atlas) db engine.

Reply:
Agreed on the non-Atlas plugin